Lib/ufo2ft/kernFeatureWriter.py

```diff
@@ -99,32 +113,139 @@ def _collectClassKerning(self):
+                        nlGlyphs.add(lGlyph)
+                        nrGlyphs.add(rGlyph)
+                        seen[pair] = val
+            nlClass, nrClass = lClass, rClass
+            if nlGlyphs != set(lGlyphs):
+                nlClass = self._liststr(sorted(nlGlyphs))
+            if nrGlyphs != set(rGlyphs):
+                nrClass = self._liststr(sorted(nrGlyphs))
+            if nlClass != lClass or nrClass != rClass:
+                self.classPairKerning[nlClass, nrClass] = val
+                del self.classPairKerning[lClass, rClass]
 
-        self._collectClassKerning()
+    def _addGlyphClasses(self, lines):
+        """Add glyph classes for the input font's groups."""
 
-        # write the glyph classes
-        lines = []
-        self._addGlyphClasses(lines)
-        lines.append("")
+        for key, members in sorted(self.leftUfoClasses.items() +
+                                   self.rightUfoClasses.items()):
+            lines.append("%s = [%s];" % (key, " ".join(members)))
 
-        # write the feature
-        lines.append("feature kern {")
-        self._addKerning(lines)
-        lines.append("    subtable;")
-        self._addKerning(lines, self.leftClassKerning, enum=True)
-        lines.append("    subtable;")
-        self._addKerning(lines, self.rightClassKerning, enum=True)
-        lines.append("    subtable;")
-        self._addKerning(lines, self.classPairKerning)
-        lines.append("} kern;")
+    def _addKerning(self, lines, kerning, enum=False):
+        """Add kerning rules for a mapping of pairs to values."""
+
+        enum = "enum " if enum else ""
+        for (left, right), val in sorted(kerning.items()):
+            lines.append("    %spos %s %s %d;" % (enum, left, right, val))
+
+    def _liststr(self, glyphs):
+        """Return string representation of a list of glyph names."""
+
+        return "[%s]" % " ".join(glyphs)
+
+    def _isClassName(self, nameRe, name):
+        """Return whether a given name matches a given class name regex."""
+
+        return re.match(nameRe, name) is not None
 
-        # return the feature, unless it's empty
-        return "" if len([ln for ln in lines if ln]) == 5 else linesep.join(lines)
+    def _makeFeaClassName(self, name):
+        """Make a glyph class name which is legal to use in OTF syntax.
+
+        Ensures the name starts with "@" and only includes characters in
+        "A-Za-z0-9._", and isn't already defined.
+        """
+
+        if not name.startswith("@"):
+            name = "@" + name
+        name = re.sub(r"[^A-Za-z0-9._]", r"", name)
+        existingClassNames = (
+            self.leftFeaClasses.keys() + self.rightFeaClasses.keys() +
+            self.leftUfoClasses.keys() + self.rightUfoClasses.keys())
+        if name in existingClassNames:
+            raise ValueError('New glyph class name "%s" (from UFO groups) '
+                             'is the name of an existing class.')
```

Comment:
Can't you mangle the name in that case? (I use basic mangling e.g. [here](https://github.com/trufont/trufont/blob/master/Lib/defconQt/fontView.py#L1516))

This code will choke when groups end up similar after dropping forbidden characters, e.g. names `my_group1` and `my_group1@` will trigger this code. 
It's not acceptable that given valid and distinct UFO group names, ufo2ft would crash.

Note that as I said in #10, since it's just making group names for compiling an FEA file, you may drop existing names entirely and generate them from a template with numbers e.g. `KERN_L_000000000001` and incrementing the number for each group from there. This way you don't need to check for duplicates and mangle until you have a valid name.